### PR TITLE
Make init mask lazy for fully initialized/uninitialized const allocations

### PR DESCRIPTION
**Pull request by lqd**
There are a few optimization opportunities in the `InitMask` and related const `Allocation`s (e.g. by taking advantage of the fact that it's a bitset that represents initialization, which is often entirely initialized or uninitialized in a single call, or gradually built up, etc).

There's a few overwrites to the same state, multiple writes in a row to the same indices, the RLE scheme for `memcpy` doesn't always compress, etc.

Here, we start with:
- avoiding materializing the bitset's blocks if the allocation is fully initialized/uninitialized
- dealloc blocks when fully overwriting, including when participating in `memcpy`s
- take care of the fixme about allocating blocks of 0s before overwriting them to the expected value
- expanding unit test coverage of the init mask

This should be most visible on benchmarks and crates where const allocations dominate the runtime (like `ctfe-stress-5` of course), but I was especially looking at the worst cases from #93215.

This first change allows the majority of `set_range` calls to stay with a lazy init mask when bootstrapping rustc (not that the init mask is a big part of the process in cpu time or memory usage).

r? @oli-obk 

I have another in-progress branch where I'll switch the singular initialized/uninitialized value to a watermark, recording the point after which everything is uninitialized. That will take care of cases where full initialization is monotonic and done in multiple steps (e.g. an array of a type without padding), which should then allow the vast majority of const allocations' init masks to stay lazy during bootstrapping (though interestingly I've seen such gradual initialization in both left-to-right and right-to-left directions, and I don't think a single watermark can handle both).

### CODE REVIEW

**Comment by lqd** [comment]
The try build completed a while ago.

@rust-timer build 89b1ad9f82f52b6fb2a2fe1ac2f139b2e8240846

File: tests/codegen/consts.rs [hunk 1]
```diff
@@ -9,7 +9,7 @@
 // CHECK: @STATIC = {{.*}}, align 4
 
 // This checks the constants from inline_enum_const
-// CHECK: @alloc_76bfe2f13a3e3b01074971d122eac57e = {{.*}}, align 2
+// CHECK: @alloc_701ed935fbda2002838d0a2cbbc171e5 = {{.*}}, align 2
```
Comment (lqd) [inline tests/codegen/consts.rs:12]: In this test and the one below, I've had to change the expected name for some allocations. 

My intuition is that the layout of masks changed and that changed the hash (if that's one) used in these symbols, is that correct @oli-obk ? I'm not familiar with how the constants get their names. 

I'm guessing that checking the full constants' symbols is not common in codegen tests, and that's why only these two needed changing.

**Comment by saethlin** [comment]
Awesome! I've been meaning to write this optimization for a while.

**Comment by saethlin** [comment]
Just for kicks, how much improvement do you get on this kind of code:
```rust
fn main() {
    for _ in 0..4 {
        let a = [0u8; 1024 * 1024 * 1024];
        drop(&a[..]);
    }
}
```

**Comment by lqd** [comment]
I remember the results from your example in #93215 I recently looked at

```rust
const ROWS: usize = 100000;
const COLS: usize = 10000;

static TWODARRAY: [[u128; COLS]; ROWS] = [[0; COLS]; ROWS];

fn main() {}
```
emitting metadata went from 17.7s to 16.1s (and from 78s to 55s with assertions turned on ^^ -- there are quite expensive ones in the init mask's bit search functions).

File: tests/codegen/consts.rs (hunk 1 above)
Comment (oli-obk) [reply tests/codegen/consts.rs:12]: yes, that is entirely expected due to the reasons you intuited.

File: compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs [hunk 2]
```diff
@@ -10,20 +10,185 @@ type Block = u64;
...
+/// value.
+#[derive(Clone, Debug, Eq, PartialEq, TyEncodable, TyDecodable, Hash, HashStable)]
 pub struct InitMask {
-    blocks: Vec<Block>,
+    pub(crate) blocks: InitMaskBlocks,
     len: Size,
 }
 
+#[derive(Clone, Debug, Eq, PartialEq, TyEncodable, TyDecodable, Hash, HashStable)]
+pub(crate) enum InitMaskBlocks {
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs:24]: Why do so many things here have to be public to the entire crate?

Comment (lqd) [reply compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs:24]: For the unit tests in the `tests` module. They look at the mask structure to count the number of actually materialized blocks. It seemed cleaner and harder to break accidentally, than to have a `#[cfg(test)]` method returning the count ? But if you'd prefer that, either is fine by me.

Comment (RalfJung) [reply compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs:24]: You can put the tests into `interpret/allocation/init_mask/tests.rs`, then they can access private fields without having to make them public.

Comment (lqd) [reply compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs]: I'm aware, but there were existing init mask tests in the other module and didn't really want to change existing things too much here. In any case if you prefer that I've pushed a commit to do so.

I'll let you notify bors if you're happy.

**Comment by RalfJung** [comment]
Yeah I think that's better, thanks!

@bors r=oli-obk

**Pull request merged**